### DISCUSSION TITLE
Conundrum: Need to execute composer as non-root user, but cannot execute as www-data and other non-root user does not have write access to kimai directory

### DISCUSSION THREAD

**Discussion by RayCulp** [kimai/kimai#3052] [On-Premise: Installation & Updates]
I recently moved Kimai from my old shared hosting to a VPS and everything is running smoothly, but I ran into an interesting problem today when I tried to update to 1.16.10.

This was the first time I tried to update on the new VPS, so composer was not yet installed. I installed it with no problems.

But when I tried to run `php /pathto/composer/composer.phar install --no-dev --optimize-autoloader`, composer warned me not to run it as root or using `sudo` because of security concerns. I then tried `su www-data`, but it said `"This account is currently not available"`. I then switched to a different non-root user, but composer complained that the kimai folder was not writable. 

The only solution I found was to temporarily `chown` the kimai folder to my non-root user, do the update, then `chown` back to www-data.

Is there a more elegant way to do this?

**Comment by kevinpapst** [comment]
You can allow a login shell for the `www-data` user (that is what the message `This account is currently not available ` is about) or you do it with your own user account. It is enough if the group is `www-data`, the owner can be someone else. 

It requires that you set the permissions of and inside the Kimai directory once with root and then switch back to your regular user account. 
Another issue will then occur: after flushing the cache, new directories were created in your user context and those are not writeable by the Webserver. And you might not be able to "fix" the permissions for the `var/` directory (AKA: change group to `www-data`) with your regular user, unless you add this account to the `www-data` group. This should allow your regular user account to change the owner of the new created cache directory to `www-data`.

You can also use a `setfacl` if it is available as pointed out here: https://symfony.com/doc/current/setup/file_permissions.html
But this needs a deeper understanding of the Linux ACL world and I would recommend to use the first method instead.
It is only the first time that you have to set it up properly, afterwards it works as if you would have used the root user.

Finally: this is not really a Kimai issue, but more a Linux hosting question. But I hope your Kimai experience just got better by the higher performance of the VPS.

**Reply by RayCulp** [reply]
Thanks for the reply. You're right, it's not really a Kimai question, but I thought it might still be relevant. 

I did some [reading](https://serverfault.com/questions/388016/bash-scripting-su-to-www-data-for-single-command) and I think I can achieve the desired result using `sudo -u www-data` and then `php /pathto/composer/composer.phar install --no-dev --optimize-autoloader`
